File: doc/code/cpp/coff.cpp
#include <iostream>
#include <memory>

#include <LIEF/COFF.hpp>

void sections() {
  // lief-doc: sections-start
  std::unique_ptr<LIEF::COFF::Binary> coff;

  for (const LIEF::COFF::Section& section : coff->sections()) {
    std::cout << section.name() << '\n';
  }
  // lief-doc: sections-end
}

void disassemble() {
  // lief-doc: disassemble-start
  std::unique_ptr<LIEF::COFF::Binary> coff;

  for (const auto& inst : coff->disassemble("?foo@@YAHHH@Z")) {
    std::cout << inst.to_string() << '\n';
  }

  // Using demangled representation
  for (const auto& inst : coff->disassemble("int __cdecl bar(int, int)")) {
    std::cout << inst.to_string() << '\n';
  }
  // lief-doc: disassemble-end
}

void parse_forms() {
  // lief-doc: parse-start

  // Using a file path as a std::string
  std::unique_ptr<LIEF::COFF::Binary> coff = LIEF::COFF::Parser::parse("test.obj");
  // lief-doc: parse-end
}

File: doc/code/rust/src/coff.rs
pub fn sections(some_coff: &lief::coff::Binary) {
    // lief-doc: sections-start
    let coff: &lief::coff::Binary = some_coff;

    for section in coff.sections() {
        println!("{section:?} {section}");
    }
    // lief-doc: sections-end
}

pub fn disassemble(some_coff: &lief::coff::Binary) {
    // lief-doc: disassemble-start
    let coff: &lief::coff::Binary = some_coff;

    for inst in coff.disassemble_function("?foo@@YAHHH@Z") {
        println!("{}", inst);
    }

    // Using demangled representation
    for inst in coff.disassemble_function("int __cdecl bar(int, int)") {
        println!("{}", inst);
    }
    // lief-doc: disassemble-end
}

pub fn parse_forms() {
    // lief-doc: parse-start
    let coff: lief::coff::Binary = lief::coff::Binary::parse("test.obj").unwrap();
    // lief-doc: parse-end
}
